File: src/helper/trim_and_clean.c
/**
* trim_and_clean.c
*
* (C) Copyright 2025 AtlantisOS Project
* by @NachtsternBuild
*
* License: GNU GENERAL PUBLIC LICENSE Version 3
*
* @brief Removes leading and trailing spaces/quotation marks from a string
* 
* Usage:
* trim_and_clean(value_buffer);
*
*/

#include "helper.h"

/**
 * @brief Removes leading and trailing spaces/quotation marks from a string
 * @param str The string to be cleaned up.
 * @return The cleaned string.
 */
char *trim_and_clean(char *str) 
{
    if (!str || *str == '\0') 
    {
        return str;
    }
    
    // remove newline and empty space
    char *end = str + strlen(str) - 1;
    while (end > str && isspace((unsigned char)*end)) 
    {
        *end-- = '\0';
    }

    // remove empty space at start
    char *start = str;
    while (*start && isspace((unsigned char)*start)) 
    {
        start++;
    }
    
    // remove '"' at start and at end
    if (start[0] == '"' && start[strlen(start) - 1] == '"') 
    {
        start++;
        start[strlen(start) - 1] = '\0';
    }

    // copy at the start of the function
    if (start != str) 
    {
        memmove(str, start, strlen(start) + 1);
    }

    return str;
}

File: src/helper/file/trim_and_clean.rs
//! Utility for cleaning up strings.
//!
//! This module provides functions to remove unnecessary spaces and 
//! enclosing quotation marks from strings, which is particularly important when parsing 
//! user input or configuration values.
/**
* trim_and_clean.rs
*
* (C) Copyright 2026 AtlantisOS Project
* by @NachtsternBuild
*
* License: GNU GENERAL PUBLIC LICENSE Version 3
*/

/// Removes leading and trailing whitespace as well as enclosing quotation marks.
///
/// The function operates efficiently on string slices (`&str`), so no 
/// new string allocations in memory are required.
///
/// # How it works
/// 1. First, all whitespace at the beginning and end is removed (`trim`).
/// 2. If the remaining text begins **and** ends with `"`, these 
///    two characters are removed.
///
/// # Arguments
/// * `s` - The string to be cleaned.
///
/// # Return Value
/// A cleaned string slice.
///
/// # Usage:
///
/// ```rust
/// fn main() {
///    let input = "  \"Hello AtlantisOS\"  ";
///    let cleaned = trim_and_clean(input);
///    println!("Input: |{}|", input);
///    println!("Output: |{}|", cleaned);
/// }
/// ```
pub fn trim_and_clean(s: &str) -> &str {
    // remove empty space
    let mut trimmed = s.trim();

    // check for '"'
    if trimmed.len() >= 2 && trimmed.starts_with('"') && trimmed.ends_with('"') {
        // remove 
        trimmed = &trimmed[1..trimmed.len() - 1];
    }

    trimmed
}
